### PR TITLE
ghc: Linux build fixes for all versions

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

This PR fixes several issues I discovered while building GHC on Linux

1. It disables NUMA, which requires `libnuma`, a dependency for which we do not have a formula.
2. It disables PDF documentation generation, which needs LaTeX, a complex dependency surely don't want to add.  Unfortunately the approach for doing this is very clunky, requiring us to create a `build.mk` file with the option in it rather than setting environment variables or passing an argument, but that is how the GHC documentation instructs users to do this, and is also how Gentoo does it.
3. It fixes the code for patching the RPATH and interpreter for the stage 0 binaries.  For some reason, `file.dynamic_elf?` doesn't actually return false for the one file which the path globbing pulls in (`ghc-cabal-bindist`), and `keg.change_rpath` fails because that file is not a dynamic ELF.  Instead simply excluding that file was sufficient to make everything work.  If anyone has a suggestion about why `file.dynamic_elf?` is not returning false for files which are not dynamic ELFs, I am happy to change how the RPATH/interpreter patching logic works.

### CODE REVIEW

**Comment by danielnachun** [comment]
Based on the discussion in https://github.com/Homebrew/linuxbrew-core/pull/21804, we can remove the code in here that depends brewed `glibc` because these formula should have `cellar :any_skip_relocation` and therefore don't need to be built from source.  We already had `cellar :any_skip_relocation` for `ghc@8.6`, but it needs to be added to `ghc@8.8` and `ghc`.  

There are two other small build fixes that I discovered which I will leave in to make life easier if someone ever does decide to build this from source.

File: Formula/ghc@8.8.rb [hunk 1]
```diff
@@ -7,11 +7,11 @@ class GhcAT88 < Formula
   revision 1
```
Comment (iMichka) [inline Formula/ghc@8.8.rb:7]: revision OS.mac? ? 1 : 2

File: Formula/ghc@8.6.rb [hunk 2]
```diff
@@ -12,7 +12,6 @@ class GhcAT86 < Formula
     sha256 "af21e24b89361083a6cd5a27268e0470cdbf2e8616d1d95355df603f58f4e30d" => :catalina
     sha256 "ccbe2725d127cc1ddd2142294fd62981d6cd7ab110f56b1faa2560c28276b822" => :mojave
     sha256 "67a54e9d669e51b8018d064b771d31079421b777b03077dc7f02949ecdf8b0c0" => :high_sierra
-    sha256 "2c7cbe7d8ce9046631bf67c6a5e3f3ded8ed41f3d3fc5bca8c0b176d4801e92a" => :x86_64_linux
```
Comment (iMichka) [inline Formula/ghc@8.6.rb:15]: Bump the revision above:
revision OS.mac? ? 2 : 3

File: Formula/ghc.rb [hunk 3]
```diff
@@ -11,11 +11,11 @@ class Ghc < Formula
   end
 
   bottle do
+    cellar :any_skip_relocation
     sha256 "36daf8d12bf15ea04e12e9df4dc8a21de5ed01b39a3fd32ef3663ddb44a459fc" => :big_sur
     sha256 "1fc15705e5b9a24ca773d2df745345e75496ec3f227ea518d9945d6e1add6e18" => :catalina
     sha256 "cf8e2755d55d2a6479e4cebbee041ff63bece9bbae7d514c0b7601f6542b8081" => :mojave
     sha256 "40e2b31e3390cf784dedcd4357362bee577b6a5c1d911f5c072828fe01f42e8f" => :high_sierra
-    sha256 "98d149359e56a7570c97f0722838851534275da7998d96d32bcf1baad5428d50" => :x86_64_linux
```
Comment (iMichka) [inline Formula/ghc.rb:18]: Bump the revision above:
revision OS.mac? ? 2 : 3

**Comment by iMichka** [comment]
Nice cleanup, looks good. Can you bump the revisions too so that the newly built bottles can be uploaded and shipped.

**Comment by danielnachun** [comment]
Thanks!  I know you mentioned in (https://github.com/Homebrew/linuxbrew-core/pull/21929) that `cellar :any` might be lost if we have another revision bump after this.  Is that a concern here?  It would obviously be a huge pain for you or another maintainer to have to manually readd `cellar :any_skip_relocation` after every revision.  If it helps I can add a comment above the bottle block to at least document that `ghc` should always be `cellar :any_skip_relocation`.

**Review by iMichka — APPROVED** [review]

**Comment by iMichka** [comment]
Let's discuss this in #21929. On the next ghc merge from Mac to the linux repo, I will keep the cellar line, that should be fine. What we need to look is what happens in the future: when there are no merges anymore, the tooling will auto-add / auto-remove this cellar information. I still need to better understand the process too, some things are buggy and need to be fixed. Having separate cellar information for each platform is a first step towards removing some weird logic around cellar lines.

**Pull request closed without merging**